File: Algoritmos/matlab/newton.m

function [raiz, iteraciones] = newton(f, df, x0, tol, max_iter)
% NEWTON Método de Newton-Raphson para encontrar raíces
%
% Sintaxis: [raiz, iteraciones] = newton(f, df, x0, tol, max_iter)
%
% Parámetros:
%   f        - Handle de la función
%   df       - Handle de la derivada
%   x0       - Aproximación inicial
%   tol      - Tolerancia (opcional, default: 1e-6)
%   max_iter - Máximo de iteraciones (opcional, default: 100)
%
% Retorna:
%   raiz        - Aproximación de la raíz
%   iteraciones - Matriz con historial [n, x, f(x), f'(x), x_nuevo, error]

    % Valores por defecto
    if nargin < 4, tol = 1e-6; end
    if nargin < 5, max_iter = 100; end

    % Inicialización
    x = x0;
    iteraciones = zeros(max_iter, 6);

    for i = 1:max_iter
        fx = f(x);
        dfx = df(x);

        % Verificar división por cero
        if abs(dfx) < 1e-10
            fprintf('Error: Derivada muy pequeña en iteración %d\n', i);
            raiz = NaN;
            iteraciones = iteraciones(1:i-1, :);
            return;
        end

        x_nuevo = x - fx / dfx;
        error = abs(x_nuevo - x);

        % Guardar iteración
        iteraciones(i, :) = [i, x, fx, dfx, x_nuevo, error];

        % Criterio de convergencia
        if error < tol
            fprintf('Convergencia alcanzada en %d iteraciones\n', i);
            raiz = x_nuevo;
            iteraciones = iteraciones(1:i, :);
            return;
        end

        x = x_nuevo;
    end

    fprintf('No se alcanzó convergencia en %d iteraciones\n', max_iter);
    raiz = x;
    iteraciones = iteraciones(1:max_iter, :);
end

% Ejemplo de uso:
% f = @(x) x^2 - 2;
% df = @(x) 2*x;
% [raiz, hist] = newton(f, df, 1.0);
% fprintf('Raíz encontrada: %.10f\n', raiz);

File: Algoritmos/python/newton.py
"""
Método de Newton-Raphson para encontrar raíces de funciones
Fórmula: x_{n+1} = x_n - f(x_n)/f'(x_n)
"""


def newton(f, df, x0, tol=1e-6, max_iter=100):
    """
    Método de Newton-Raphson

    Parámetros:
    -----------
    f : function
        Función de la cual se busca la raíz
    df : function
        Derivada de la función f
    x0 : float
        Aproximación inicial
    tol : float
        Tolerancia para el criterio de parada
    max_iter : int
        Número máximo de iteraciones

    Retorna:
    --------
    raiz : float
        Aproximación de la raíz
    iteraciones : list
        Historial de iteraciones
    """
    iteraciones = []
    x = x0

    for i in range(max_iter):
        fx = f(x)
        dfx = df(x)

        # Verificar división por cero
        if abs(dfx) < 1e-10:
            print(f"Error: Derivada muy pequeña en iteración {i}")
            return None, iteraciones

        x_nuevo = x - fx / dfx

        iteraciones.append({
            'n': i,
            'x': x,
            'f(x)': fx,
            "f'(x)": dfx,
            'x_nuevo': x_nuevo,
            'error': abs(x_nuevo - x)
        })

        # Criterio de convergencia
        if abs(x_nuevo - x) < tol:
            print(f"Convergencia alcanzada en {i + 1} iteraciones")
            return x_nuevo, iteraciones

        x = x_nuevo

    print(f"No se alcanzó convergencia en {max_iter} iteraciones")
    return x, iteraciones


# Ejemplo de uso
if __name__ == "__main__":
    import math

    # Ejemplo: Encontrar raíz de f(x) = x^2 - 2 (raíz de 2)
    f = lambda x: x ** 2 - 2
    df = lambda x: 2 * x

    raiz, hist = newton(f, df, x0=1.0)

    if raiz:
        print(f"\nRaíz encontrada: {raiz:.10f}")
        print(f"Verificación f(raíz) = {f(raiz):.2e}")

        print("\nHistorial de iteraciones:")
        for it in hist:
            print(f"n={it['n']}: x={it['x']:.6f}, f(x)={it['f(x)']:.6f}, error={it['error']:.2e}")
